### PR TITLE
Feat/precompiled contracts loadtest

### PR DESCRIPTION
**Pull request by IdrisHanafi**
# Description

Currently `polycli` only supports loadtesting with the standard EVM opcodes. This PR introduces loadtesting with EVM [precompiled contracts](https://www.evm.codes/precompiled?fork=merge). 

Another thing to note with the precompiled contracts is `testECAdd` (address `0x06`) and `testECMul` (address `0x07`) fail when running via loadtester, and needs further digging. For now, I've commented it out from being ran randomly under `GetRandomPrecompiledContractAddress()`. You can force it to run by passing `--mode a --function 6` (or function 7 for ecMul) to the loadtester. 

__Other Minor Changes:__
Also updated the function `testDIFFICULTY(uint x) public returns(uint)`, in `contracts/LoadTester.sol`, to reflect [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399) changes. Specifically, changing the `difficulty()` function to `prevrandao()`. The side effect of this change will only be in compiling `LoadTester.sol` requiring a solidity compiler (`solc`) version of 0.8.18+ or later (more notes [here](https://docs.soliditylang.org/en/v0.8.19/yul.html#evm-dialect)). You can `brew install solidity` for the latest version.

## Jira / Linear Tickets

- [DVT-615](https://polygon.atlassian.net/browse/DVT-615)
- [DVT-636](https://polygon.atlassian.net/browse/DVT-636)

# Testing
Run geth locally with exact params of:
```
$ geth --dev --dev.period 2 --http --verbosity 5 --rpc.gascap 5000000000 --miner.gaslimit 5000000000 --dev.gaslimit 5000000000
```

Fund loadtest account:
```
$ geth attach "http://127.0.0.1:8545"
> eth.sendTransaction({from: eth.coinbase, to: "0x85da99c8a7c2c95964c8efd687e95e632fc533d6", value: web3.toWei(50000000000, "ether")})
```

Run loadtester:
Run random precompiled addresses with `--mode p`:
```
$ go run main.go loadtest http://localhost:8545 --verbosity 600 --concurrency 1 --requests 10 --rate-limit 0 --mode p --chain-id 1337 --summarize
...
1:00PM DBG Starting main loadtest loop currentNonce=1
1:00PM TRC Starting Thread routine=0
1:00PM TRC Finished starting go routines. Waiting..
1:00PM TRC Executing contract method method=TestBlake2f
1:00PM TRC Request mode=p nonce=1 request=0 routine=0
1:00PM TRC Executing contract method method=TestECPairing
1:00PM TRC Request mode=p nonce=2 request=1 routine=0
1:00PM TRC Executing contract method method=TestModExp
1:00PM TRC Request mode=p nonce=3 request=2 routine=0
1:00PM TRC Executing contract method method=TestECRecover
1:00PM TRC Request mode=p nonce=4 request=3 routine=0
1:00PM TRC Executing contract method method=TestECPairing
1:00PM TRC Request mode=p nonce=5 request=4 routine=0
1:00PM TRC Executing contract method method=TestECPairing
1:00PM TRC Request mode=p nonce=6 request=5 routine=0
1:00PM TRC Executing contract method method=TestIdentity
1:00PM TRC Request mode=p nonce=7 request=6 routine=0
1:00PM TRC Executing contract method method=TestBlake2f
1:00PM TRC Request mode=p nonce=8 request=7 routine=0
1:00PM TRC Executing contract method method=TestModExp
1:00PM TRC Request mode=p nonce=9 request=8 routine=0
1:00PM TRC Executing contract method method=TestModExp
1:00PM TRC Request mode=p nonce=10 request=9 routine=0
1:00PM DBG Finished main loadtest loop currentNonce=11
1:00PM DBG Waiting for transactions to actually be mined
1:00PM TRC Not all transactions have been mined. Waiting currentNonce=1 endNonce=11
1:00PM TRC It looks like all transactions have been mined currentNonce=11 endblock=86 startblock=84
1:00PM INF rough test summary (ignores errors) firstBlockTime=2023-03-30T13:00:35-04:00 lastBlockTime=2023-03-30T13:00:39-04:00 testDuration=4 tps=2.5 transactionCount=10
1:00PM TRC Starting block range capture
1:00PM TRC Fetching block range end=86 start=84
1:00PM TRC Starting tx receipt capture
1:00PM TRC Fetching tx receipt range end=1 start=0 startblock=0x54
1:00PM INF Fetched tx receipts hashes=1 receipts=1
1:00PM TRC Fetching tx receipt range end=10 start=0 startblock=0x56
1:00PM INF Fetched tx receipts hashes=10 receipts=10
1:00PM INF Block summary len=3
1:00PM TRC Raw receipts len=11
1:00PM INF Receipt summary len=11
1:00PM DBG Encountered transaction with more than 2 hours latency blockNumber=84 mineTime=2023-03-30T13:00:35-04:00 nonce=0 requestTime=0001-01-01T00:00:00Z txHours=2562047.7880152157
Block number: 86        Time: 2023-03-30 13:00:39 -0400 EDT     Gas Limit: 5,000,000,000        Gas Used: 832,325       Num Tx: 10      Utilization 0%  Latencies: 3.428        3.539   3.686
Successful Tx: 10       Total Tx: 10
Total Mining Time: 2s
Total Transactions: 10
Total Gas Used: 832,325
Transactions per sec: 5
Gas Per Second: 416,162.5
Latencies - Min: 3.428  Median: 3.539   Max: 3.686
1:00PM TRC Total Summary Time summaryTime=3.026833ms
1:00PM INF * Results
1:00PM INF Samples samples=10
1:00PM INF Start startTime=2023-03-30T13:00:35-04:00
1:00PM INF End endTime=2023-03-30T13:00:35-04:00
1:00PM INF Mean Wait meanWait=0.026199099999999996
1:00PM INF Num errors numErrors=0
1:00PM INF Finished
```

Run specific contract address with `--mode a --function 1` (this example is running `0x01` aka `ecRecover`):
```
$ polygon-cli git:(feat/precompiled-contracts-loadtest) go run main.go loadtest http://localhost:8545 --verbosity 600 --concurrency 1 --requests 10 --rate-limit 0 --mode a --function 1 --chain-id 1337 --summarize
...
12:56PM DBG Starting main loadtest loop currentNonce=456
12:56PM TRC Starting Thread routine=0
12:56PM TRC Finished starting go routines. Waiting..
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=456 request=0 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=457 request=1 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=458 request=2 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=459 request=3 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=460 request=4 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=461 request=5 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=462 request=6 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=463 request=7 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=464 request=8 routine=0
12:56PM TRC Executing contract method method=TestECRecover
12:56PM TRC Request mode=a nonce=465 request=9 routine=0
12:56PM DBG Finished main loadtest loop currentNonce=466
12:56PM DBG Waiting for transactions to actually be mined
12:56PM TRC Not all transactions have been mined. Waiting currentNonce=456 endNonce=466
12:56PM TRC It looks like all transactions have been mined currentNonce=466 endblock=4654 startblock=4652
12:56PM INF rough test summary (ignores errors) firstBlockTime=2023-03-30T12:56:52-04:00 lastBlockTime=2023-03-30T12:56:56-04:00 testDuration=4 tps=2.5 transactionCount=10
12:56PM TRC Starting block range capture
12:56PM TRC Fetching block range end=4654 start=4652
12:56PM TRC Starting tx receipt capture
12:56PM TRC Fetching tx receipt range end=1 start=0 startblock=0x122c
12:56PM INF Fetched tx receipts hashes=1 receipts=1
12:56PM TRC Fetching tx receipt range end=10 start=0 startblock=0x122e
12:56PM INF Fetched tx receipts hashes=10 receipts=10
12:56PM INF Block summary len=3
12:56PM TRC Raw receipts len=11
12:56PM INF Receipt summary len=11
12:56PM DBG Encountered transaction with more than 2 hours latency blockNumber=4652 mineTime=2023-03-30T12:56:52-04:00 nonce=455 requestTime=0001-01-01T00:00:00Z txHours=2562047.7880152157
Block number: 4,654     Time: 2023-03-30 12:56:56 -0400 EDT     Gas Limit: 5,000,000,000        Gas Used: 275,050       Num Tx: 10      Utilization 0%  Latencies: 3.272        3.307   3.338
Successful Tx: 10       Total Tx: 10
Total Mining Time: 2s
Total Transactions: 10
Total Gas Used: 275,050
Transactions per sec: 5
Gas Per Second: 137,525
Latencies - Min: 3.272  Median: 3.307   Max: 3.338
12:56PM TRC Total Summary Time summaryTime=24.72525ms
12:56PM INF * Results
12:56PM INF Samples samples=10
12:56PM INF Start startTime=2023-03-30T12:56:52-04:00
12:56PM INF End endTime=2023-03-30T12:56:52-04:00
12:56PM INF Mean Wait meanWait=0.00690735
12:56PM INF Num errors numErrors=0
12:56PM INF Finished
```

- [X] Run random precompiled contracts
- [X] Run specific precompiled contracts

[DVT-615]: https://polygon.atlassian.net/browse/DVT-615?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/loadtest/loadtest.go [hunk 1]
```diff
@@ -62,16 +62,18 @@ import (
...
+	loadTestModeDeploy               = "d"
+	loadTestModeCall                 = "c"
+	loadTestModeFunction             = "f"
+	loadTestModeInc                  = "i"
+	loadTestModeRandom               = "r"
+	loadTestModeStore                = "s"
+	loadTestModeLong                 = "l"
+	loadTestModeERC20                = "2"
+	loadTestModeERC721               = "7"
+	loadTestModePrecompiledContracts = "p"
```
Comment (gatsbyz) [inline cmd/loadtest/loadtest.go:75]: sweet

File: contracts/contracts.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	// "github.com/ethereum/go-ethereum/core/vm"
```
Comment (rebelArtists) [inline contracts/contracts.go:13]: nit: remove

File: contracts/precompiledContracts.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+		// 7, // NOTE: ecMul requires a lot of gas and buggy
+		8,
+		9,
+	}
+
+	return codes[rand.Intn(len(codes))]
+	/**
+	n := 9
+	return rand.Intn(n) + 1 // [1, n + 1)
+	*/
```
[the remark this answers is not in the document]
Comment (IdrisHanafi) [reply contracts/precompiledContracts.go]: updated!

File: cmd/loadtest/loadtest.go [hunk 4]
```diff
@@ -532,7 +538,7 @@ func mainLoop(ctx context.Context, c *ethclient.Client, rpc *ethrpc.Client) erro
 	// deploy and instantiate the load tester contract
 	var ltAddr ethcommon.Address
 	var ltContract *contracts.LoadTester
-	if strings.ContainsAny(mode, "rcfisl") || *inputLoadTestParams.ForceContractDeploy {
+	if strings.ContainsAny(mode, "rcfislpas") || *inputLoadTestParams.ForceContractDeploy {
```
Comment (praetoriansentry) [inline cmd/loadtest/loadtest.go:541]: We'll need a more manageable way to do this at some point haha.

File: cmd/loadtest/loadtest.go [hunk 5]
```diff
@@ -692,7 +698,7 @@ func mainLoop(ctx context.Context, c *ethclient.Client, rpc *ethrpc.Client) erro
 		return err
 	}
 	startNonce := currentNonce
-	log.Debug().Uint64("currenNonce", currentNonce).Msg("Starting main loadtest loop")
+	log.Debug().Uint64("currentNonce", currentNonce).Msg("Starting main loadtest loop")
```
Comment (praetoriansentry) [inline cmd/loadtest/loadtest.go:701]: nice catch...

File: cmd/loadtest/loadtest.go [hunk 6]
```diff
@@ -918,6 +930,34 @@ func loadtestCall(ctx context.Context, c *ethclient.Client, nonce uint64, ltCont
...
+
+func loadtestCallPrecompiledContracts(ctx context.Context, c *ethclient.Client, nonce uint64, ltContract *contracts.LoadTester, useSelectedAddress bool) (t1 time.Time, t2 time.Time, err error) {
+	var f int
+	ltp := inputLoadTestParams
+
+	chainID := new(big.Int).SetUint64(*ltp.ChainID)
+	privateKey := ltp.ECDSAPrivateKey
+	iterations := ltp.Iterations
+	if useSelectedAddress {
+		f = int(*ltp.Function)
```
Comment (praetoriansentry) [inline cmd/loadtest/loadtest.go:942]: Is the idea here that you would pass `9` in order to call `blake2f`

File: contracts/LoadTester.sol [hunk 7]
```diff
@@ -561,18 +561,23 @@ contract LoadTester {
...
         uint result = 0xDEADBEEF0044;
         assembly {
             let v := 0
             for { let i := 0 } lt(i, x) { i := add(i, 1) }
             {
-                v := difficulty()
+                // NOTE: Post Paris, the difficulty() instruction has been supplanted for prevrandao() 
+                // Source: https://eips.ethereum.org/EIPS/eip-4399
+                // v := difficulty()
+                v := prevrandao()
```
Comment (praetoriansentry) [inline contracts/LoadTester.sol:575]: Very interesting... This is a weird refactor tbh. I.e. it's still named `DIFFICULTY` but it looks like it's still `0x44` opcode. I think this if okay for now but we should talk about changing the name.

File: contracts/precompiledContracts.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+	"github.com/ethereum/go-ethereum/common/math"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/rs/zerolog/log"
+)
+
+func GenerateECRecoverInput(privateKey *ecdsa.PrivateKey) []byte {
+	message := []byte("Test ecRecover")
+	messageHash := crypto.Keccak256Hash(message)
```
Comment (praetoriansentry) [inline contracts/precompiledContracts.go:22]: One thing I would like to be able to test here are empty values. In the future maybe we can have a table with some test values. see https://github.com/0xPolygon/polygon-edge/pull/1260/files#diff-f9f7ba4c9a0e5e981794760f2c0bebe923604444b918224398d14a16d89b2172R150

**Review by praetoriansentry — APPROVED** [review]
I think it's fine to merge and we can make some future improvements.

File: contracts/precompiledContracts.go (hunk 8 above)
Comment (IdrisHanafi) [reply contracts/precompiledContracts.go:22]: +1, I really like the idea of using different load testing inputs. Tracking for future effort here: [DVT-642](https://polygon.atlassian.net/browse/DVT-642)

File: cmd/loadtest/loadtest.go (hunk 6 above)
Comment (IdrisHanafi) [reply cmd/loadtest/loadtest.go:942]: That's correct, `ltp.Function` is multipurpose for an opcode and a precompiled address.

File: cmd/loadtest/loadtest.go (hunk 5 above)
Comment (IdrisHanafi) [reply cmd/loadtest/loadtest.go:701]: breaking change lol

**Pull request merged**